File: app/src/main/java/com/example/rickmortymvvm/util/adapter/DiffUtilGeneric.java
package com.example.rickmortymvvm.util.adapter;

import androidx.recyclerview.widget.DiffUtil;

import java.util.List;

public class DiffUtilGeneric<T extends DiffUtilGeneric.Compare<T>> extends DiffUtil.Callback {
    List<T> oldList;
    List<T> newList;

    public DiffUtilGeneric(List<T> oldList, List<T> newList) {
        this.oldList = oldList;
        this.newList = newList;
    }

    @Override
    public int getOldListSize() {
        return oldList != null ? oldList.size() : 0;
    }

    @Override
    public int getNewListSize() {
        return newList != null ? newList.size() : 0;
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return newList.get(newItemPosition).getId() == oldList.get(oldItemPosition).getId();
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return newList.get(newItemPosition).isEqual(oldList.get(oldItemPosition));
    }

    public interface Compare<T> {
        int getId();
        boolean isEqual(T o);
    }
}

File: app/src/main/java/com/example/rickmortymvvm/app/util/adapter/DiffUtilGeneric.kt
package com.example.rickmortymvvm.app.util.adapter

import androidx.recyclerview.widget.DiffUtil
import com.example.rickmortymvvm.app.util.adapter.DiffUtilGeneric.Compare

class DiffUtilGeneric<T : Compare<in T>>(var oldList: List<T>, var newList: List<T>) : DiffUtil.Callback() {
    override fun getOldListSize(): Int {
        return  oldList.size
    }

    override fun getNewListSize(): Int {
        return newList.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return newList[newItemPosition].id == oldList[oldItemPosition].id
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return newList[newItemPosition].isEqual(oldList[oldItemPosition])
    }

    interface Compare<T> {
        val id: Int
        fun isEqual(o: T): Boolean
    }
}
